scripts/list-tests-from-{json,robot}.sh: add new scripts

krystian-hebel opened the pull request:
Can be used to check whether test cases in source files and their copy in `test_cases.json` are in sync.

krystian-hebel committed:
scripts/list-tests-from-{json,robot}.sh: add new scripts

Can be used to check whether test cases in source files and their
copy in `test_cases.json` are in sync.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel commented:
PR marked as a draft since the task includes fixes to the test cases, but scripts and docs can already be reviewed.

krystian-hebel committed:
docs/adding-and-naming-test-cases.md: fix environment-test-ids name

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

philipanda commented on docs/adding-and-naming-test-cases.md:
```diff
@@ -66,6 +65,10 @@ DSP002.201 - External HDMI display in OS (Ubuntu)
 
+If the ID is changed again for the same test, the newer `Previous ID` is added
+to the end of line, with single space as a separator (i.e. no comma or semicolon
+between them). This format is expected by a script described [later](#validate).
```
An example is worth 1000 words

krystian-hebel replied:
https://github.com/Dasharo/open-source-firmware-validation/pull/876/commits/a910152d3489fb765b38b5db6368b45a07b160b7

philipanda commented on docs/adding-and-naming-test-cases.md:
```diff
@@ -158,6 +161,50 @@ existing one by pointing to the new ID.
+  (i.e. test cases in all Dasharo `*.robot` files), and prints a sorted list of
+  all of them.
+- `scripts/list-tests-from-json.sh` - finds active (without `changed_to` field)
+  and obsolete (without that field) IDs in `test_cases.json` file, and prints a
```
```suggestion
  and obsolete (with that field) IDs in `test_cases.json` file, and prints a
```

krystian-hebel replied:
https://github.com/Dasharo/open-source-firmware-validation/pull/876/commits/a910152d3489fb765b38b5db6368b45a07b160b7

philipanda commented on docs/adding-and-naming-test-cases.md:
```diff
@@ -158,6 +161,50 @@ existing one by pointing to the new ID.
+- `scripts/list-tests-from-json.sh` - finds active (without `changed_to` field)
+  and obsolete (without that field) IDs in `test_cases.json` file, and prints a
+  list of all of them. The script doesn't sort the output, which indirectly
+  checks that JSON file is properly sorted.
```
So maybe the script should check if that's the case?

krystian-hebel replied:
`Possible use cases` do that, by comparing with always sorted list from `.robot` files. This script is for printing only, and the checks are to be run by the caller. If (when) they are used by CI, then it would be checked there, possibly with another tests. Right now (at least for me) the more common use case is _showing_ the differences, not just checking if there are any.

philipanda commented on scripts/list-tests-from-robot.sh:
```diff
@@ -0,0 +1,22 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# Get lines that start with ID, store each line as array element
+mapfile -t active < <(grep -Rh "^[A-Z]\{1,6\}[0-9]\{3\}\.[0-9]\{3\}" dasharo-*)
```
ETHPERF has 7 letters and will be left out by this regular expression.

What if someone creates an invalid test case, like one containing 4 numbers after the dot, and forgets to add it to the json? That case won't be catched.

I think the expressions shouldn't be that restrictive when it comes to the number of characters

krystian-hebel replied:
Nice catch. I was wondering why I only saw `DEPRECATED` tests there, but I wrongfully assumed that they were moved to NET.

I've changed it to less restrictive `^[A-Z]*[0-9]\{1,10\}\.[0-9]\{1,10\}` and checked that it doesn't give any false-positives: https://github.com/Dasharo/open-source-firmware-validation/pull/876/commits/7c9ba52aabda497c5d0519679459959bb1c2494e

krystian-hebel committed:
docs/adding-and-naming-test-cases.md: add example for Previous IDs

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
scripts/list-tests-from-robot.sh: relax ID search pattern

Previous pattern missed ETHPERF. It would also miss cases with too
many digits - there aren't such cases in OSFV now, but there were in
previous releases.

Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>
